### DISCUSSION TITLE
On demand revalidation with webhook on specific path

### DISCUSSION THREAD

**Discussion by ZarckDev** [vercel/next.js#38258] [Help]
Hi all,

_I'm using the Next.js version 12.2.0, and using GraphCMS as headless cms_

I'm making a blog with webhook triggers and On demand functionnality.
Revalidation is perfectly working on the `/blog` path but having issues on specific path like `/blog/[slug]`
**I get a 'failed to revalidate' specific path** on Vercel Logs, 404 error when the post did not exist on the first build and 500 when I try to delete a post that was there when I built the project.
I followed the https://github.com/leerob/on-demand-isr project and they seem to do the same thing I do.

Here is my webhook code (graphcms is sending a body, with the concerned post slug):
`/api/webhook.ts`
```js
import type { NextApiRequest, NextApiResponse } from 'next';
const { verifyWebhookSignature } = require('@graphcms/utils');

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).send('Only POST request allowed');
  }

  const body = req.body;

  console.log('[Next.js] Revalidating /blog');
  await res.revalidate('/blog');
  if (body.data.slug) {
    console.log(`[Next.js] Revalidating /blog/${body.data.slug}`);  // I checked, the path with slug is correct on logs
    await res.revalidate(`/blog/${body.data.slug}`);
  }

  return res.json({ revalidated: true, slug: body.data.slug });
}
```

data fetching `/blog/[slug].tsx`
```js
export const getStaticPaths: GetStaticPaths = async () => {
  const posts = (await getAllSlugs()) || [];

  return {
    paths: posts.map(({ slug }: Params) => ({ params: { slug } })),
    fallback: false,
  };
};

export const getStaticProps: GetStaticProps<Props, Params> = async ({
  params,
}) => {
  const post = (await getPostDetails(params!.slug)) || [];

  return {
    props: {
      post,
    },
  };
};
```

data fetching `/blog/index.tsx`
```js
export const getStaticProps: GetStaticProps = async () => {
  const posts = (await getAllPosts()) || [];

  return {
    props: {
      posts,
    },
  };
};
```

I guess all of this is because the path of the slug does not exists initially. And revalidation does not support the deletion of the path.

Am I missing something here on what the `revalidate` function does?
I guess we have to update all paths on get getStaticPaths, but how to?

Thank you very much!

**Comment by icyJoseph** [comment]
I wonder if you could try with, `unstable_onlyGenerated`, set to true:

```ts
 revalidate: (urlPath: string, opts?: {
        unstable_onlyGenerated?: boolean;
    }) => Promise<void>;
```

For deleted posts, and such, I think your `getPostDetails` should still return something, that your UI can use to render a fallback, or a notice that the post is gone.

You could also try/catch `getPostDetails`, and just return `{ notFound: true, revalidate: 100}` when it fails to fetch a post because it was deleted. 

It is important to include a revalidate because **otherwise the page is cached, by the browser, for a very long time as a 404.**

```ts
export const getStaticProps: GetStaticProps<Props, Params> = async ({
  params,
}) => {
  try {
    const post = (await getPostDetails(params!.slug)) || [];

    return {
      props: {
        post,
      },
    };
  } catch (e) {
    // perhaps analyse here why it failed...
    return { notFound: true, revalidate: 100 };
  }
};
```


Please do test out the `unstable_onlyGenerated` option, I wanted to setup an example, but ran out of time today. I'll try it tomorrow, since I actually don't lnow what this flag does 😅😂

**Reply by ZarckDev** [reply]
Hi!

Thank you for this quick answer and also for your suggestions.

I tried this `unstable_onlyGenerated` options for `revalidate`. 
```js
{ .... }
console.log('[Next.js] Revalidating /blog');
await res.revalidate('/blog', {
  unstable_onlyGenerated: true,
});
if (body.data.slug) {
  console.log(`[Next.js] Revalidating /blog/${body.data.slug}`);
  await res.revalidate(`/blog/${body.data.slug}`, {
    unstable_onlyGenerated: true,
  });
}
return res.json({ revalidated: true, slug: body.data.slug });
{ .... }
```
I don't know what it's really doing but it does not throw anymore 404 or 500 errors on `/blog/${body.data.slug}` revalidation, for post deleted or added. 
Revalidation on my `/blog` index page using webhook are still working by the way.

But the behaviour is the same, my blog specific paths are not updated, whether on delete or add from CMS.

Thank you also for you suggestion on the try...catch, I've added it to my getStaticProps but no errors popping...

I don't really know what to do next.

**Comment by shykhov** [comment]
Hi @icyJoseph, thank you for your advice! I've got the same problem and as for @ZarckDev `unstable_onlyGenerated` helped me to omit 404, but a specific dynamic blog page still throw 404

by the way
`return { notFound: true, revalidate: 100 }`
Did not resolve the problem on the page refresh (checked in incognito mode also)

Is there anything we can do / miss?

**Comment by shykhov** [comment]
![Screenshot 2022-09-12 at 17 16 07](https://user-images.githubusercontent.com/17176104/189677241-f8a44a4f-6615-4550-8341-a7b032d2cad1.png)
My problem was in `fallback: false`

@ZarckDev it executes `getStaticPath` only at a building process, so the page does not exists, try `fallback: 'blocking'` instead
